### DISCUSSION TITLE
Recreate file asset object hash from client side

### DISCUSSION THREAD

**Discussion by fouss** [directus/directus#16226] [Archive]
I tried to recreate the hash of the image transformation to be able to get transformed image directly from my CDN.
The idea is to generate all the images transformations during the generation of the project and set the image src of the CDN instead of the directus url with the transformation.
I use object-hash exactly as it's done [here](https://github.dev/directus/directus/blob/main/api/src/services/assets.ts), when I run dev script on nuxt, everything is fine

Problem:
during the "nuxt generate", the result of the hash are not the same with the exact same parameters.
I finally find why [with this issue of object-hash](https://github.com/puleos/object-hash/issues/62). With `respectType: false` as option on the `hash` function, i have the same result !
BUT it's not the same as the one generated by default by directus [here](https://github.com/directus/directus/blob/main/api/src/services/assets.ts#L183) because the call is done in browser.
Adding the option the function will fix thix but it will "invalidate" all the existing transformations. What do you think about it ?

### To Reproduce

run `hash("anything")` from a browser and `hash("anything")` from node
fix:
run `hash("anything", {respectType: true})` from a browser and `hash("anything", {respectType: true})` from node

**Comment by SeanDylanGoff** [comment]
Out of interest:
What's the use case for getting the files directly? Because if you're using a proxying CDN like CloudFlare (?) the file requests should only rarely hit directus anyways, as the CDN is gonna cache the files.

**Comment by fouss** [comment]
I am using the CDN azure, the difference for the same image is 1.4s to get the image from directus (url with transformations) and 6ms when I hit directly the blob by the CDN

**Comment by SeanDylanGoff** [comment]
that's horrible! do normal API requests (or /server/ping) also take so long? Our Instances only take that long on /files when converting an unsuitable input on first request (15MP PNG for example)

And where are the files stored?

**Comment by rijkvanzanten** [comment]
That definitely sounds like the files aren't actually being cached in the CDN then. The assets endpoint will perform permissions checks, so it'll be slightly slower than the "raw" access (as that has no permissions at all), but really shouldn't be counted in seconds when the file already exists. 

In any case, I'll move this to a discussion as the OP isn't really a bug in Directus, but a usage question for things outside of it 👍🏻

**Comment by fouss** [comment]
The thing is, I cannot have http2 for directus and I cannot handle the cache as I want because of my infra. The possibility to allow object-hash to generate the same hash on every environment with the same data (I think that should be the case by default) by setting an environment variable seems to be a pretty good option to me, It will allow anyone to do what he wants with his transformed images.

**Comment by tmaihoff** [comment]
I am facing the same problem.

Accessing the image via the Directus assets endpoint results in request times of up to 6 ! seconds while it takes about 25ms directly from the CDN. This behaviour is crosschecked via the Directus logs, Postman tests and actual behaviour in the browser.

@fouss have you found a solution for generating the hash on the client side?
Or better: a solution for the slow Directus API?

**Reply by rijkvanzanten** [reply]
I'd recommend putting the CDN in front of Directus, rather than accessing it directly. The file hash suffix is generated with `object-hash` from the array of transformations applied. You might be able to re-generate using the logic here: https://github.com/directus/directus/blob/47b9824954fd6538406d1ae754ec99f4744e5cff/api/src/services/assets.ts#L204-L207 though I would strongly recommend against going that route, as the transformation may or may not exist yet (Directus generates thumbnails on-demand rather than up-front)
